pandas/core/dtypes/concat.py

```diff
@@ -100,45 +96,54 @@ def concat_compat(
     # Creating an empty array directly is tempting, but the winnings would be
     # marginal given that it would still require shape & dtype calculation and
     # np.concatenate which has them both implemented is compiled.
+    orig = to_concat
     non_empties = [x for x in to_concat if _is_nonempty(x, axis)]
     if non_empties and axis == 0 and not ea_compat_axis:
         # ea_compat_axis see GH#39574
         to_concat = non_empties
 
-    dtypes = {obj.dtype for obj in to_concat}
-    kinds = {obj.dtype.kind for obj in to_concat}
-    contains_datetime = any(
-        isinstance(dtype, (np.dtype, DatetimeTZDtype)) and dtype.kind in "mM"
-        for dtype in dtypes
-    ) or any(isinstance(obj, ABCExtensionArray) and obj.ndim > 1 for obj in to_concat)
+    any_ea, kinds, target_dtype = _get_result_dtype(to_concat, non_empties)
```

Comment:
looks like the only reason to pass `non_empties` to `_get_result_dtype` is to calculate `len(non_empties)`. If so, perhaps just pass `len(non_empties) == 0` and make the second argument of `_get_result_dtype` `all_empty`?

Reply:
either way i guess

Reply:
sure, but here `to_concat` and `non_empties` could be the same thing, so it looks a bit odd to pass the same thing twice

Reply:
fair enough.  changing that in this PR will make rebasing the other one more painful.  mind if i make the change in that one after this goes in?

Reply:
sure